### DISCUSSION TITLE
How can libvips read the image file from local disk to get the memory data of the image for direct rendering like QT QImage?

### DISCUSSION THREAD

**Discussion by cool2528** [libvips/libvips#2726] [Q&A] [answered]
Hello I would like to ask why the data I save using `vips_jpegsave_buffer ` is not the data after the image has been loaded? Because I can't use this memory data for ffmpeg, but when I use QT's QImage to load the image, it is indeed decompressed data that can be used directly for ffmpeg's frame->data, so how can I properly use livips to load the image from the local file and then get the memory data of this file like QT's QImage, which is the decompressed image data that can be rendered directly?

This code makes me to get the data that can be rendered directly from QImage with the help of QT's secondary. Can libvips read the image data that can be rendered directly? Instead of compressed file memory stream data?
```
 VipsImage* image = vips_image_new_from_file(afilePath.c_str(), "access", VIPS_ACCESS_LAST, nullptr);
            if (!image) return nResult;
            srcFrame = av_frame_alloc();
            if (!srcFrame)
            {
                g_object_unref(image);
                return nResult;
            }
            width = vips_image_get_width(image);
            height = vips_image_get_height(image);
            srcFrame->format = AV_PIX_FMT_RGB24;
            srcFrame->width = width;
            srcFrame->height = height;
            size_t len = 0;
            uint8_t* pBuffer = nullptr;
            if (vips_jpegsave_buffer(image, (void**)&pBuffer, &len, "interlace",true,nullptr) < 0)
            {
                av_frame_free(&srcFrame);
                g_object_unref(image);
                return nResult;
            }
            QImage img = QImage::fromData(pBuffer, len);
            if (img.format() != QImage::Format_RGB888)
            {
                img = img.convertToFormat(QImage::Format_RGB888);
            }
            int nsize = av_image_fill_arrays(srcFrame->data, srcFrame->linesize, img.bits(), AV_PIX_FMT_RGB24, width, height, 1);
            if (transfrom(srcFrame, frame))
            {
                nResult = 0;
            }
            g_free(pBuffer);
            frame->pts = pts;
            g_object_unref(image);
```

**Comment by jcupitt** [comment]
Hello @cool2528,

You can get a pointer to the RGBRGB etc. C-style array of pixel values like this:

```C
VipsImage *image = vips_image_new_from_file("something.jpg", NULL);
const void *data = vips_image_get_data( image );
```

The memory is owned by `image`, so you need to keep `image` alive for as long as you use the pointer.

Use `width`, `height`, `bands` (often called channels, it'll be 3 for RGB, 4 for RGBA etc.), `format` (ie. numeric type, perhaps `uchar`) and `interpretation` (eg. `srgb` for RGB and RGBA, or `cmyk` for CMYK images) to determine how to transform the values for display.

After calling `vips_image_get_data()` you can use the macro `VIPS_IMAGE_ADDR()` to get a pointer to a pixel. But you must call `vips_image_get_data()` first to force the image into memory.

If you know your image is 8-bit RGB (perhaps a JPEG image?) you can do:

```C
unsigned char *pixel = (unsigned char*) VIPS_IMAGE_ADDR( image, 10, 10 );
```

To get a pointer to the pixel at (10, 10).

See: 

https://www.libvips.org/API/current/libvips-header.html#vips-image-get-data

https://www.libvips.org/API/current/VipsImage.html#VIPS-IMAGE-ADDR:CAPS

**Comment by jcupitt** [comment, marked as answer]
... however, although this is very easy, it will not work well for large images. The call to `get_data()` can take a long time, and it can allocate large temporary files.

libvips has a special display sink you can connect to the end of a pipeline. You can use user navigation in the window to drive evaluation of the image processing pipeline, it'll all be threaded and asynchronous, and it will often be able to display an image directly from the source file, with no decompression stage.

There's a tiny demo program here to show how it works (300 lines of C). It uses gtk, but it should be simple to adapt to Qt.

https://github.com/libvips/vipsdisp-tiny

And here's a full image viewer done in the same style:

https://flathub.org/apps/details/org.libvips.vipsdisp

https://github.com/jcupitt/vipsdisp

That one uses the GPU to render the image: tiles are streamed in from libvips in the background and used to make a set of textures arranged in a pyramid of scales, then as you navigate around the image, it builds a scene graph of the viewable region and renders that to the display.

**Reply by jcupitt** [reply]
`vipsdisp` has a secret key -- press 'd' to put the renderer into debug mode. It'll draw all the top-level tiles in the pyramid with an age. If you zoom quickly, you'll see tiles swapping in and out as they come from disc. If you pick one of the slower visualization modes (log scale + false colour, perhaps) it's clearer.

**Reply by cool2528** [reply]
Thank you, thank you for your answer this is very helpful

**Reply by cool2528** [reply]
@jcupitt Thank you, thank you for your answer this is very helpful

**Comment by wccxigua** [comment]
> Hello I would like to ask why the data I save using `vips_jpegsave_buffer ` is not the data after the image has been loaded? Because I can't use this memory data for ffmpeg, but when I use QT's QImage to load the image, it is indeed decompressed data that can be used directly for ffmpeg's frame->data, so how can I properly use livips to load the image from the local file and then get the memory data of this file like QT's QImage, which is the decompressed image data that can be rendered directly?
> 
> This code makes me to get the data that can be rendered directly from QImage with the help of QT's secondary. Can libvips read the image data that can be rendered directly? Instead of compressed file memory stream data?
> 
> ```
>  VipsImage* image = vips_image_new_from_file(afilePath.c_str(), "access", VIPS_ACCESS_LAST, nullptr);
>             if (!image) return nResult;
>             srcFrame = av_frame_alloc();
>             if (!srcFrame)
>             {
>                 g_object_unref(image);
>                 return nResult;
>             }
>             width = vips_image_get_width(image);
>             height = vips_image_get_height(image);
>             srcFrame->format = AV_PIX_FMT_RGB24;
>             srcFrame->width = width;
>             srcFrame->height = height;
>             size_t len = 0;
>             uint8_t* pBuffer = nullptr;
>             if (vips_jpegsave_buffer(image, (void**)&pBuffer, &len, "interlace",true,nullptr) < 0)
>             {
>                 av_frame_free(&srcFrame);
>                 g_object_unref(image);
>                 return nResult;
>             }
>             QImage img = QImage::fromData(pBuffer, len);
>             if (img.format() != QImage::Format_RGB888)
>             {
>                 img = img.convertToFormat(QImage::Format_RGB888);
>             }
>             int nsize = av_image_fill_arrays(srcFrame->data, srcFrame->linesize, img.bits(), AV_PIX_FMT_RGB24, width, height, 1);
>             if (transfrom(srcFrame, frame))
>             {
>                 nResult = 0;
>             }
>             g_free(pBuffer);
>             frame->pts = pts;
>             g_object_unref(image);
> ```

Hi @cool2528 ,Can you tell me how i use Vips in QT-MSVC? Thanks very much!!!

**Reply by cool2528** [reply]
Hi @wccxigua You can download a compiled version of libvips from here
https://github.com/libvips/build-win64-mxe/releases
You can bring in the corresponding lib library file in visual studio and use it

**Reply by wccxigua** [reply]
Thanks for your help !  I've actually tried many times to use the libs,but seems i can't use it ,I don't know what the problem is,or can you tell me how to use it correctly  =_=

![image](https://user-images.githubusercontent.com/35681295/159209470-ad062e35-9b85-4741-9469-1bd67c403a3f.png)

**Reply by cool2528** [reply]
Hi, @wccxigua you are using the libvips c++ code here, as far as I know if you want to use the libvips c++ version under the MSVC compiler you need to compile the c++ version of the wrapper yourself https://github.com/libvips/libvips/tree/master/ cplusplus 
Have a look here because https://github.com/libvips/build-win64-mxe is cross-compiled in linux using mingw GCC 
I suggest you use the libvips C SDK if you don't want to compile it yourself.
Just bring in
`#include "vips/vips.h"` and that's it

**Reply by wccxigua** [reply]
> 

Thanks again!I will operate it as you suggest.

**Reply by jcupitt** [reply]
Exactly, C++ doesn't have an ABI, so you can't mix C++ compilers in one project.
